net/src/client/http/mod.rs

```diff
@@ -0,0 +1,351 @@
+        address: String,
+        /// An error message.
+        msg: String,
+    },
+    /// Found an unknown HTTP status code.
+    #[fail(
+        display = "Unknown HTTP status code ({}). Underlying error: {}",
+        code, msg
+    )]
+    UnknownStatusCode {
+        /// The unknown status code.
+        code: u16,
+        /// An error message.
+        msg: String,
+    },
+    /// Found an unknown HTTP version.
+    #[fail(display = "Unknown HTTP version ({})", version)]
+    UnknownVersion {
+        /// The unknown HTTP version.
+        version: String,
+    },
+    /// Tried to process an HTTP request with an unsupported HTTP method.
+    #[fail(
+        display = "Tried to process an HTTP request with an unsupported HTTP method {}",
+        method
+    )]
+    UnsupportedMethod {
+        /// The unsupported HTTP method.
+        method: String,
+    },
+}
+
+impl WitnetHttpClient {
+    /// Create a new `WitnetHttpClient`
+    pub fn new(proxy: &Option<String>) -> Result<Self, WitnetHttpError> {
+        // Try to parse the provided proxy address, if any, and fail if the address is not a valid URI
+        let proxy_uri = if let Some(address) = proxy {
+            Some(isahc::http::Uri::from_str(address.as_ref()).map_err(|err| {
+                WitnetHttpError::InvalidProxyUri {
+                    address: address.clone(),
+                    msg: err.to_string(),
+                }
+            })?)
+        } else {
+            None
+        };
+
+        // Build an `isahc::HttpClient`. Will use the proxy URI, if any
+        let client = Arc::new(
+            isahc::HttpClient::builder()
+                .proxy(proxy_uri)
+                .build()
+                .map_err(|err| WitnetHttpError::ClientBuildError {
+                    msg: err.to_string(),
+                })?,
+        );
+
+        Ok(Self { client })
+    }
+
+    /// Turn this `WitnetHttpClient` into a `surf`-compatible client
+    pub fn as_surf_client(&self) -> surf::Client {
+        surf::Client::with_http_client(self.clone())
+    }
+}
+
+impl Default for WitnetHttpClient {
+    fn default() -> Self {
+        Self::new(&None).unwrap()
+    }
+}
+
+/// Enables interoperability between `isahc::Request` and `surf::http::Request`.
+pub struct WitnetHttpRequest {
+    req: isahc::Request<isahc::AsyncBody>,
+}
+
+impl From<isahc::Request<isahc::AsyncBody>> for WitnetHttpRequest {
+    fn from(req: isahc::Request<isahc::AsyncBody>) -> Self {
+        Self { req }
+    }
+}
+
+impl TryFrom<&mut surf::http::Request> for WitnetHttpRequest {
+    type Error = WitnetHttpError;
+
+    fn try_from(req: &mut surf::http::Request) -> Result<Self, Self::Error> {
+        let method = isahc::http::Method::from(WitnetHttpMethod::try_from(req.method())?);
+        let version = req
+            .version()
+            .ok_or(WitnetHttpError::UnknownVersion {
+                version: String::from("None"),
+            })
```

Comment:
This error is actually ignored because of the `unwrap_or_default` below. Maybe this entire method chain can be refactored into a function

    fn(Option<surf::http::Version>) -> isahc::http::Version

to make the intention to ignore errors more clear.